### ISSUE TITLE
Numpy binary incompatibility warning

### ISSUE THREAD

**Issue by hassec** [Unidata/netcdf4-python#1354]
To create a minimal reproducer (I used Python 3.11.9):

```
python -m venv test_nc4
source test_nc4/bin/activate
pip install netcdf4
python -c "import numpy;import warnings;warnings.filterwarnings('error');import netCDF4"
```

This will raise an error with: 
`numpy.ndarray size changed, may indicate binary incompatibility. Expected 16 from C header, got 96 from PyObject`

This isn't numpy 2.0 specific, `pip install netcdf4 "numpy<2.0"` will raise the same warning.

To the best of my understanding this means that the numpy that was used during the `netcdf4` build is incompatible with the numpy version I'm using. 

I know that this is usually just a warning, but this means I've got to manually ignore these in my pytest suite.

It's also hard to gauge for a user whether this warning is problematic. Thus I'm wondering if this can be fixed on the netcdf4 side?

**Comment by mhogan-nwra** [comment]
On my Windows and Linux platforms, this is not a warning but an actual ValueError 

```language=shell
>>> import netCDF4
Traceback (most recent call last):
  File "C:\Users\mhogan\AppData\Local\pypoetry\Cache\virtualenvs\apex-n-5SkYZI-py3.9\lib\site-packages\IPython\core\interactiveshell.py", line 3550, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-9588a3d4fb24>", line 1, in <module>
    import netCDF4
  File "C:\Users\mhogan\AppData\Local\Programs\PyCharm Professional\plugins\python-ce\helpers\pydev\_pydev_bundle\pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "C:\Users\mhogan\AppData\Local\pypoetry\Cache\virtualenvs\apex-n-5SkYZI-py3.9\lib\site-packages\netCDF4\__init__.py", line 31, in <module>
    from ._netCDF4 import *
  File "C:\Users\mhogan\AppData\Local\Programs\PyCharm Professional\plugins\python-ce\helpers\pydev\_pydev_bundle\pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "src\\netCDF4\\_netCDF4.pyx", line 1, in init netCDF4._netCDF4
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
>>> np.__version__
Out[3]: '2.0.2'
```

The error disappears when I downgrade to NumPy 1.26.4

**Platform specifics**

Windows 11.0
Memory: 8192M
Cores: 20
CPython version: 3.9.13

**Comment by ChrisBarker-NOAA** [comment]
I've been seeing this warning a lot -- unfortunately, I can't tell where it's coming from. I use netCDF4 a lot, but I also use all kinds of other stuff that builds against numpy, so I'm not sure it's netCDF4's fault.

But I use conda-forge for everything, so in that case, *something* is getting built and pinned wriong in conda-forge. 

Which is odd, as conda forge does have numpy version pinning.

But PyPi packages are more of a problem here.

Hmm -- now that I think about it, I've had those working before numpy2, so ???

Anyway, maybe not useful but one more data point.

**Comment by djhoese** [comment]
As another data point, I only see this when installing netcdf4-python 1.6.x. Installing 1.7.x seems to all be built with numpy 2.x so backwards compatibility makes it work with numpy 1.x (I think). My unrelated problem now is that installing packages from PyPI results in not being able to use compression plugins (ex. `import hdf5plugin` has no effect).

**Comment by ocefpaf** [comment]
@hassec what is the Python, netcdf4, and numpy versions in your vitualenv? Note that, it may be fine, regardless of the warning.

@ChrisBarker-NOAA you should not see those in conda-forge. If you do, please report in the netcdf4-feedstock.

@djhoese can you please open a new issue for the plugin problem? It seems unrelated to the discussions here.

**Comment by djhoese** [comment]
I should have noted I only had issues with PyPI packages. Letting conda install the newest compatible conda-forge packages worked as expected.

**Comment by ocefpaf** [comment]
@hassec I was able to reproduce that warning, but the test suit passed when installing that wheel locally. 

We did build those wheels with `numpy>=2.0.0rc1` for `python >=3.9`, and it does break if you downgrade numpy, indicating that the ABI is compatible with 2.0. My guess is that the warning can be ignored.

You can see [from the logs](https://productionresultssa18.blob.core.windows.net/actions-results/7860fa73-ae9b-4b65-98ab-2aac034af034/workflow-job-run-f0914f7d-c81c-514c-3006-0cb1124193ef/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-12-16T15%3A37%3A48Z&sig=qI6mlM8NdFfGWNTWd9BcPv7DVOiLfM4pg8Dyvemc6vg%3D&ske=2024-12-17T03%3A20%3A37Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-12-16T15%3A20%3A37Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-11-04&sp=r&spr=https&sr=b&st=2024-12-16T15%3A27%3A43Z&sv=2024-11-04) that, for python 3.12, the numpy used was 2.1.2.

2024-10-22T17:49:10.7521563Z Collecting numpy (from netCDF4==1.7.2)
2024-10-22T17:49:10.7534725Z   Using cached numpy-2.1.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)

Maybe something change from 2.1 to 2.2 that trigger that warning? Sill, they should be compatible.

**Comment by hassec** [comment]
I can reproduce with numpy 2.1.2 as well: <img width="948" alt="Image" src="https://github.com/user-attachments/assets/c4edec81-8ab6-4ccc-bdfc-087dd1008ef9" />

> ... they should be compatible 
 
That's fine for me for a one-off usecase, but this is exactly the kind of uncertainty that I don't want to put on a user, so I've been avoiding using this library for now, as these warnings make users uneasy.

**Comment by ocefpaf** [comment]
If warnings make your users uneasy, it will be hard for them to use any Software.

BTW, numpy is the one silencing those warnings:

https://github.com/numpy/numpy/blob/main/numpy/__init__.py#L320-L323

A lot of the history behind that can be read in https://github.com/numpy/numpy/issues/11788

TL;DR I don't see what we can do here to solve this. It is unlikely a problem with netCDF4.

Your options are:

1. Use the conda package, my guess is that the Cython used there does't trigger these;
2. Pay someone to build your own wheels in your own controlled environment;
3. Accept that the warnings is just a harmless warning and use the wheels that the community built.

**Comment by ChrisBarker-NOAA** [comment]
> @ChrisBarker-NOAA you should not see those in conda-forge. If you do, please report in the netcdf4-feedstock.

I was getting those a lot in the numpy-2 transition, not so much anymore. And we have not only netCDF4, but a bunch of other code that's compiled against numpy, so I was never sure exactly what was triggering it.

So: nothing to be done here  -- it/when it pops up again, I'll try to take a closer look.

**Comment by hassec** [comment]
> BTW, numpy is the one silencing those warnings:
>
> https://github.com/numpy/numpy/blob/main/numpy/__init__.py#L320-L323

Thanks, I was not aware of this. I'd argue then that the problem is actually the user code that was doing `warnings.filterwarnings('error')` and thus unknowingly overwriting the numpy ignores.

Let's close this, I agree that it doesn't seem to be a netcdf issue.

**Comment by veenstrajelmer** [comment]
> BTW, numpy is the one silencing those warnings:
> https://github.com/numpy/numpy/blob/main/numpy/__init__.py#L320-L323

Here is a permalink so people can still find it in the future (and also today the line numbers have changed, but I assume you meant these lines): https://github.com/numpy/numpy/blob/46268caa719db95ab2198bcd619f0e3120d7c52e/numpy/__init__.py#L704-L707

**Comment by cssherman** [comment]
I ran into a similar issue when writing a unit test that uses netCDF4.  Running this small test with pytest:

```python
import xarray as xr

def test_write_netcdf():
    ds = xr.Dataset(
        {'foo': (('x'), [0, 1])},
        coords={
            'x': [0, 1]
        },
    )
    ds.to_netcdf('example.nc', engine='netcdf4')
```

Produces the following warning:

```
tests/test_import.py::test_write_netcdf
  <frozen importlib._bootstrap>:241: RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 16 from C header, got 96 from PyObject
```

As others have suggested, importing netCDF4 first suppresses the warning.  Moving the xarray import into scope of _test_write_netcdf_ (or adding "ignore:numpy.ndarray size changed:RuntimeWarning" to the pytest filterwarnings) also seems to work here.  I'm seeing this issue with both a pip install  (netCDF4==1.7.4, numpy==2.3.4, xarray==2026.1.0) and with a fresh conda install (netCDF4==1.7.4, numpy==2.4.4, xarray==2026.2.0).

**Comment by ChrisBarker-NOAA** [comment]
```
  <frozen importlib._bootstrap>:491: RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 16 from C header, got 96 from PyObject
```

That makes me nervous -- this warning does seem to indicate that somthing is out of wack -- though it's probably a build issue, rather than a code issue -- i.e. a conda-forge issue.

For my part, I was seeing these a lot, and then not for a while, and now I'm seeing them again :-(  

I posted an issue on the netCDF4 feedstock:

https://github.com/conda-forge/netcdf4-feedstock/issues/182

The good news is, I've never had an actual problem with it at runtime ...

**Comment by ChrisBarker-NOAA** [comment]
OK -- with the discussion in the feedstock, it became clear that it was not a conda-forge build issue.

numpy code that silences the warning:

> https://github.com/numpy/numpy/blob/main/numpy/__init__.py#L320-L323

```
    # Filter out Cython harmless warnings
    warnings.filterwarnings("ignore", message="numpy.dtype size changed")
    warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
    warnings.filterwarnings("ignore", message="numpy.ndarray size changed")
```

> Thanks, I was not aware of this. I'd argue then that the problem is actually the user code that was doing warnings.filterwarnings('error') and thus unknowingly overwriting the numpy ignores.

In my case, and probably others, I think it's pytest that is overloading the warnings filter. pytest, by default, shows you all the warnings. Which is a good thing, in general.

The trick here is that apparently even numpy thinks it's a useless warning, so it really should never be raised at all -- apparently an issue for Cython:

https://github.com/cython/cython/issues/7533

But back to netCDF4.

It seem there may be a way for it not to trigger this quite so much, e.g. from the numpy code:

`ctypedef class numpy.ndarray [object PyArrayObject, check_size ignore]`

as found by @xylar in: 

https://github.com/conda-forge/netcdf4-feedstock/pull/183

(that's a patch to the netCDF4 code, that can be applied when building for conda-forge)

**Comment by fgoudreault** [comment]
+1 I recently stumbled onto the same issue as @cssherman albeit not using condaforge but a plain python 3.12 venv with numpy and netcdf-python latest versions. The warning only occurs when using pytest and, only if the netcdf library is not imported inside a test function for the first time. For instance, if I import netCDF4 outside the test function, (plain `import netCDF4` at the top of the file with no direct usage), then the warning disappears.

**Comment by da-woods** [comment]
Just to quickly copy the information I found from the Cython issue (I'm pretty sure this is mostly stuff that already been figured out in other places):

The variation in size is probably due to:

```
#if !defined(NPY_NO_DEPRECATED_API) || \
    (NPY_NO_DEPRECATED_API < NPY_1_7_API_VERSION)
```

defined in the Numpy headers. And maybe also because of mismatched Numpy versions. But mainly that define.

As Chris said yesterday, you can ignore the warning by changing

https://github.com/Unidata/netcdf4-python/blob/25c1684eae440d733f1ea05ffdeb386b2a9ae559/include/netCDF4.pxi#L376-L377

to

```
    ctypedef extern class numpy.ndarray [object PyArrayObject, check_size ignore]:
        pass
```

(I'm expecting some future version of Cython though will notice that this class is empty and ignore the warning more generally, unless you try to inherit from it).

---------------------------------------

Note that this is only safe because Numpy have gone to some effort to make it safe by keeping binary compatibility. If you see a similar error from a non-Numpy extension type then it's probably a real problem and you shouldn't just ignore it.

**Comment by ChrisBarker-NOAA** [comment]
Thanks @da-woods !

I've done a PR for this: #1471 

If it all builds and tests pass, it should be an easy update.
